File: 609-find-duplicate-file-in-system/find-duplicate-file-in-system.js
/**
 * @param {string[]} paths
 * @return {string[][]}
 */
// Time & Space: O(n * l)...l is the longest path in paths
var findDuplicate = function(paths) {
    let groups = new Map()
    for (let path of paths) {
        let arr = path.split(" ")
        for (let i = 1; i < arr.length; i++) {
            let file = arr[i], value = "", prefix = ""
            let leftIdx = 0
            while (file[leftIdx] !== "(") {
                prefix += file[leftIdx]
                leftIdx++
            }
            leftIdx++
            while (file[leftIdx] !== ")") {
                value += file[leftIdx]
                leftIdx++
            }
            
            if (!groups.has(value)) groups.set(value, [])
            groups.get(value).push(arr[0] + "/" + prefix)
        }
    }
    let result = []
    for (let group of Array.from(groups.values())) {
        if (group.length > 1) {
            result.push(group)
        }
    }
    return result
};

File: 609-find-duplicate-file-in-system/find-duplicate-file-in-system.swift
class Solution {
    func findDuplicate(_ paths: [String]) -> [[String]] {
        var groups:[String:[String]] = [:]
        for path: String in paths {
            let arr = path.components(separatedBy: " ")
            for i in 1..<arr.count {
                var key: String = "", p: String = arr[0] + "/", leftIdx: String.Index = arr[i].startIndex
            
                while arr[i][leftIdx] != "(" {
                    p += String(arr[i][leftIdx])
                    leftIdx = arr[i].index(after: leftIdx)
                }

                leftIdx = arr[i].index(after: leftIdx)
                while arr[i][leftIdx] != ")" {
                    key += String(arr[i][leftIdx])
                    leftIdx = arr[i].index(after: leftIdx)
                }
                groups[key, default: []].append(p)
            }
            
        }
        return Array(groups.values).filter{$0.count > 1}
    }
}